lib/livebook/runtime/evaluator.ex

```diff
@@ -449,22 +560,127 @@ defmodule Livebook.Runtime.Evaluator do
 
-    unchanged_binding =
-      Enum.filter(prev_binding, fn {key, prev_val} ->
-        val = binding_map[key]
-        :erts_debug.same(val, prev_val)
-      end)
+    identifiers_used =
+      for var_name <- vars_used(context, tracer_info, prev_context),
+          do: {:variable, var_name},
+          into: identifiers_used
+
+    identifiers_used =
+      for var_name <- tracer_info.undefined_vars,
+          do: {:variable, var_name},
+          into: identifiers_used
+
+    identifiers_defined =
+      for var_name <- vars_defined(context, prev_context),
+          do: {{:variable, var_name}, context.id},
+          into: identifiers_defined
+
+    # Modules
+
+    identifiers_used =
+      for module <- tracer_info.modules_used,
+          do: {:module, module},
+          into: identifiers_used
+
+    identifiers_defined =
+      for {module, {version, _vars}} <- tracer_info.modules_defined,
+          do: {{:module, module}, version},
+          into: identifiers_defined
+
+    # Aliases
+
+    identifiers_used =
+      for alias <- tracer_info.aliases_used,
+          do: {:alias, alias},
+          into: identifiers_used
+
+    identifiers_defined =
+      for {as, alias} <- tracer_info.aliases_defined,
+          do: {{:alias, as}, alias},
+          into: identifiers_defined
+
+    # Requires
+
+    identifiers_used =
+      for module <- tracer_info.requires_used,
+          do: {:require, module},
+          into: identifiers_used
+
+    identifiers_defined =
+      for module <- tracer_info.requires_defined,
+          do: {{:require, module}, :ok},
+          into: identifiers_defined
+
+    # Imports
+
+    identifiers_used =
+      if tracer_info.imports_used? or tracer_info.imports_defined? do
+        # Imports are not always incremental, due to :except, so if
+        # we define imports, we also implicitly rely on prior imports
+        MapSet.put(identifiers_used, :imports)
+      else
+        identifiers_used
+      end
+
+    identifiers_defined =
+      if tracer_info.imports_defined? do
+        version = :erlang.phash2({context.env.functions, context.env.macros})
+        put_in(identifiers_defined[:imports], version)
+      else
+        identifiers_defined
+      end
+
+    # Process dictionary
+
+    # Every evaluation depends on the pdict
+    identifiers_used = MapSet.put(identifiers_used, :pdict)
 
-    unchanged_binding
-    |> Enum.reduce(binding_map, fn {key, _}, acc -> Map.delete(acc, key) end)
-    |> Map.to_list()
-    |> Kernel.++(unchanged_binding)
+    identifiers_defined =
+      if context.pdict == prev_context.pdict do
+        identifiers_defined
+      else
+        version = :erlang.phash2(context.pdict)
+        put_in(identifiers_defined[:pdict], version)
+      end
+
+    {MapSet.to_list(identifiers_used), identifiers_defined}
+  end
+
+  defp vars_used(context, tracer_info, prev_context) do
+    prev_vars =
+      for {{name, nil}, _version} <- prev_context.env.versioned_vars,
```

Comment:
Wouldn't it be better to get all variables instead of filtering by `nil`? Someone could call `var!(foo, :bar)`.

Reply:
I'm actually missing a test for this, but if we include other context then we will have false positives for most macros, for example:

```elixir
Integer.is_even(2)
```

defines `{:arg1, Elixir}`, so it ends up in the binding. Then in another cell:

```elixir
Integer.is_even(2)
```

it also defines `{:arg1, Elixir}`, so it's not pruned. Pruning doesn't differentiate if the var is used or not, so we would assume it's used, meaning that the second cell depends on the first one.

Reply:
Those variables have a counter, so they should not leak beyond their original quote. I will investigate.

Reply:
Ah, the issue is particular with `defguard` I believe.

Reply:
Fixed in v1.14 and main.